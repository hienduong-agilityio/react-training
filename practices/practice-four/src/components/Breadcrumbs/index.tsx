// Libraries
import { Link } from 'react-router-dom';
import { memo } from 'react';

// SVG
import homeIcon from '@public/images/homeIcon.svg';
import arrowRight from '@public/images/arrowRightIcon.svg';

// Constants
import { ROUTE } from '@/constants';

export type BreadcrumbItem = {
  // id: The breadcrumb Items id
  id?: number;
  // label: The label to display for the breadcrumb item
  label?: string;
  // href: Optional URL for the breadcrumb item
  href?: string;
};

interface IBreadcrumbsProps {
  // items: Array of breadcrumb items
  items?: BreadcrumbItem[];
}

/**
 * Breadcrumbs component
 *
 * @returns {JSX.Element} Breadcrumbs element
 */
const Breadcrumbs = ({ items = [{ label: 'Home', href: ROUTE.ROOT }] }: IBreadcrumbsProps): JSX.Element => {
  return (
    <nav aria-label='breadcrumb'>
      <ul className='flex items-center space-x-1 md:space-x-2 rtl:space-x-reverse'>
        <li className='inline-flex items-center'>
          <Link to={ROUTE.ROOT} className='flex items-center text-sm font-medium text-gray-400 hover:text-primary-500'>
            <img src={homeIcon} alt='home' className='w-4 h-4 me-2.5 p-0' />
          </Link>
        </li>
        {items.map((item) => (
          <li key={item.label || item.id} className='inline-flex items-center'>
            <img src={arrowRight} alt='Arrow Right' className='rtl:rotate-180 w-6 h-6 text-gray-400 mx-1' />
            {item.href ? (
              <Link to={item.href} className='ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2'>
                {item.label}
              </Link>
            ) : (
              <span className='ms-1 text-sm font-medium text-gray-400 md:ms-2'>{item.label}</span>
            )}
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default memo(Breadcrumbs);
