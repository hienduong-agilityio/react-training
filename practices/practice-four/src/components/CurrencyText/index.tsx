export interface ICurrencyTextProps extends React.HTMLAttributes<HTMLParagraphElement> {
  // Currency: Optional estimated cost of the project, displayed as a financial estimate.
  currency?: number;
}

/**
 * CurrencyText component
 *
 * @returns {JSX.Element} - CurrencyItem element
 */
export const CurrencyText = ({ currency = 0, className = '' }: ICurrencyTextProps): JSX.Element => {
  return <p className={`${className}`}>{currency ? `US$ ${currency}k` : '-'}</p>;
};
