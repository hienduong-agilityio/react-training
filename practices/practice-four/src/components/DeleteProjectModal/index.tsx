// Components
import Modal from '@/components/common/Modal';
import Button from '@/components/common/Button';
import Spinner from '@/components/common/Spinner';

// Enums
import { BUTTON_COLORS, BUTTON_VARIANTS } from '@/enums/theme';

// Hooks
import { useProject } from '@/hooks/useProject';

// Store
import { ToastStore } from '@/stores/toastStore';

// Types
import type { IProjectItemProps } from '@/interfaces';

export interface IDeleteProjectModalProps {
  // projectId: The project id.
  projectId: string;
  // isModalOpen: A boolean indicating whether the modal is open.
  isModalOpen: boolean;
  // onCloseModal: A function to be called when the modal is requested to be closed.
  onCloseModal: () => void;
}

/**
 * The custom DeleteProjectModal component
 *
 * @returns {JSX.Element} - The DeleteProjectModal element
 */
export const DeleteProjectModal = ({
  projectId,
  isModalOpen = false,
  onCloseModal = () => {}
}: IDeleteProjectModalProps): JSX.Element => {
  const project = useProject({ id: projectId || '' });
  const { showToast } = ToastStore();

  const {
    mutate: { deleteProject },
    data: projectDetail,
    isMutating
  } = project;

  const projectName = (projectDetail as IProjectItemProps)?.projectName || 'this project';

  const handleDeleteProjectSuccess = () => {
    showToast('Project Deleted successfully', 'success');
    onCloseModal();
  };

  const handleDeleteProjectError = () => {
    showToast('Failed to deleted project', 'error');
    onCloseModal();
  };

  const handleDelete = () => {
    if (projectId) {
      deleteProject(projectId, {
        onSuccess: handleDeleteProjectSuccess,
        onError: handleDeleteProjectError
      });
    }
  };

  return (
    <Modal isOpen={isModalOpen} onClose={onCloseModal}>
      <div className='bg-light rounded-t-xl p-4 w-[440px]'>
        <p className='text-lg font-semibold text-gray-900 mb-3'>Delete Project</p>
        <p className='leading-6 text-gray-600'>
          Are you sure you want to delete <strong>{projectName}</strong>? If you delete, it will be permanently lost.
        </p>
      </div>
      <div className='bg-light rounded-b-xl flex justify-end gap-4 border-t-2 py-3'>
        <Button onClick={onCloseModal} variant={BUTTON_VARIANTS.OUTLINED}>
          Cancel
        </Button>
        <Button
          onClick={handleDelete}
          disabled={isMutating}
          variant={BUTTON_VARIANTS.CONTAINED}
          color={BUTTON_COLORS.DANGER}
          customClasses='mr-5'
        >
          {isMutating ? <Spinner /> : 'Delete'}
        </Button>
      </div>
    </Modal>
  );
};
