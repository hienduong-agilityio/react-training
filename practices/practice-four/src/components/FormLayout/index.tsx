interface FormLayoutProps {
  title: string;
  subtitle?: string;
  children: React.ReactNode;
}

export const FormLayout = ({ title, subtitle, children }: FormLayoutProps) => (
  <section>
    <div className='flex rounded-t-xl items-center bg-white justify-between p-4'>
      <p className='text-xl font-semibold text-gray-900'>{title}</p>
    </div>
    {subtitle && <p className='text-gray-600 mt-1 mb-4'>{subtitle}</p>}
    {children}
  </section>
);
