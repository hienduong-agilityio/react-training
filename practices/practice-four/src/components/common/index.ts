export { default as Avatar } from './Avatar';
export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Modal } from './Modal';
export { default as Pagination } from './Pagination';
export { default as SelectField } from './SelectField';
export * from './SkeletonLoader';
export { default as Spinner } from './Spinner';
export { default as StatusLabel } from './StatusLabel';
export { default as Tag } from './Tag';
export { default as Toast } from './ToastMessage';
