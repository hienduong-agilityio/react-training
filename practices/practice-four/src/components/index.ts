export { default as Breadcrumbs } from './Breadcrumbs';
export * from './CurrencyText';
export * from './DeleteProjectModal';
export * from './ErrorBoundary';
export * from './ErrorDisplay';
export * from './ProjectDetailSkeleton';
export { default as ProjectFilterDropdown } from './ProjectFilterDropdown';
export * from './ProjectFilterDropdown';
export { default as ProjectItem } from './ProjectItem';
export * from './ProjectItem';
export * from './ProjectOptionsDropdown';
export * from './ProjectTable';
export { default as ResourceTag } from './ResourcesTag';
export { default as SearchBox } from './SearchBox';
export * from './SuspenseLoader';
export { default as Timeline } from './Timeline';
export * from './ToastContainer';
export * from './FormLayout';
export * from './ProjectFormFields';
export * from './SubmitForm';

// Common
export * from './common';
