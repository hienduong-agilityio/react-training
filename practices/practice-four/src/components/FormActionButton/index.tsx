// Libraries
import { useFormStatus } from 'react-dom';

// Components
import { Button, Spinner } from '@/components';

// Enums
import { BUTTON_COLORS, BUTTON_SIZES, BUTTON_VARIANTS } from '@/enums';

interface IFormActionButtonProps {
  children: React.ReactNode;
  type?: 'submit' | 'button';
  color?: BUTTON_COLORS;
  variant?: BUTTON_VARIANTS;
  size?: BUTTON_SIZES;
  disabledOnPending?: boolean;
  showSpinner?: boolean;
  onClick?: () => void;
  customClass?: string;
}

export const FormActionButton = ({
  children,
  type = 'submit',
  variant = BUTTON_VARIANTS.CONTAINED,
  size = BUTTON_SIZES.LARGE,
  color = BUTTON_COLORS.PRIMARY,
  disabledOnPending = true,
  showSpinner = true,
  customClass = '',
  onClick
}: IFormActionButtonProps) => {
  const { pending } = useFormStatus();

  return (
    <Button
      type={type}
      disabled={disabledOnPending && pending}
      variant={variant}
      size={size}
      color={color}
      onClick={onClick}
      customClasses={customClass}
    >
      {pending && showSpinner ? <Spinner /> : children}
    </Button>
  );
};
