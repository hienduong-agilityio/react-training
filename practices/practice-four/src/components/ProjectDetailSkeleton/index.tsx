import { SkeletonLoader } from '@/components';

/**
 * ProjectDetailSkeleton components
 */
export const ProjectDetailSkeleton = () => {
  return (
    <div className='flex items-center p-8 h-full gap-36 space-y-6 flex-row'>
      <div className='flex-shrink-0 w-5/12 h-full flex flex-col items-center'>
        {/* Placeholder for Avatar */}
        <SkeletonLoader width='75%' height='300px' customClassName='mb-4' />
        <SkeletonLoader width='50%' height='20px' />
      </div>

      <div className='flex-grow lg:w-7/12 h-full space-y-6'>
        <SkeletonLoader width='70%' height='30px' customClassName='mb-4' />
        <SkeletonLoader width='20%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='30px' customClassName='mb-4 rounded-md' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='60%' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='60%' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='20px' />
      </div>
    </div>
  );
};
