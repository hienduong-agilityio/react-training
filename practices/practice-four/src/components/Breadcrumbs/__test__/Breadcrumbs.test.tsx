// Libraries
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

// Components
import { Breadcrumbs } from '@/components';

// Types
import type { BreadcrumbItem } from '@/components/Breadcrumbs';

// Constants
import { ROUTE } from '@/constants';

// Mock icons
jest.mock('@public/images/homeIcon.svg', () => 'homeIcon.svg');
jest.mock('@public/images/arrowRightIcon.svg', () => 'arrowRightIcon.svg');

describe('Breadcrumbs Component', () => {
  const defaultItems: BreadcrumbItem[] = [{ label: 'Home', href: ROUTE.ROOT }];
  const homeIconSrc = 'homeIcon.svg'; // mock icon path

  const renderBreadcrumbs = (items: BreadcrumbItem[] = defaultItems) =>
    render(
      <Router>
        <Breadcrumbs items={items} />
      </Router>
    );

  it('renders breadcrumbs with default "Home" item', () => {
    renderBreadcrumbs(defaultItems);
    const homeElement = screen.getByText('Home');

    expect(homeElement).toBeInTheDocument();
    expect(homeElement).toHaveClass('text-sm font-medium text-gray-400 hover:text-primary-500');
    expect(homeElement.closest('a')).toHaveAttribute('href', ROUTE.ROOT); // Ensuring the correct link
  });

  it('renders the correct href for an item with label and href', () => {
    const items: BreadcrumbItem[] = [{ label: 'Home', href: ROUTE.ROOT }];
    renderBreadcrumbs(items);

    const homeElement = screen.getByText('Home');
    expect(homeElement).toBeInTheDocument();
    expect(homeElement.closest('a')).toHaveAttribute('href', ROUTE.ROOT); // Ensuring the correct link
  });

  it('renders breadcrumbs with additional items and handles items without href', () => {
    const items: BreadcrumbItem[] = [
      { label: 'Home', href: ROUTE.ROOT },
      { label: 'Projects', href: ROUTE.PROJECT },
      { label: 'Project Details' }
    ];

    renderBreadcrumbs(items);

    const homeElement = screen.getByText('Home');
    const projectsElement = screen.getByText('Projects');
    const projectDetailsElement = screen.getByText('Project Details');

    // Check if the elements are rendered
    expect(homeElement).toBeInTheDocument();
    expect(projectsElement).toBeInTheDocument();
    expect(projectDetailsElement).toBeInTheDocument();

    // Check if the links are correctly applied
    expect(homeElement.closest('a')).toHaveAttribute('href', ROUTE.ROOT);
    expect(projectsElement.closest('a')).toHaveAttribute('href', ROUTE.PROJECT);
    expect(projectDetailsElement.closest('a')).toBeNull(); // No link for the last breadcrumb
  });

  it('uses id as the key when label is not provided', () => {
    const items: BreadcrumbItem[] = [{ id: 1, href: ROUTE.ROOT }]; // Item with only id
    const { container } = renderBreadcrumbs(items);

    // Verify that an element exists in the DOM and it's using id as the key.
    const listItem = container.querySelector('li');
    expect(listItem).toBeInTheDocument();
  });

  it('renders arrow icon between breadcrumb items', () => {
    const items: BreadcrumbItem[] = [
      { label: 'Home', href: ROUTE.ROOT },
      { label: 'Projects', href: ROUTE.PROJECT }
    ];

    renderBreadcrumbs(items);

    const arrows = screen.getAllByAltText('Arrow Right');
    arrows.forEach((arrow) => {
      expect(arrow).toBeInTheDocument();
      expect(arrow).toHaveClass('rtl:rotate-180 w-6 h-6 text-gray-400 mx-1');
    });
  });

  it('renders home icon with the correct attributes', () => {
    renderBreadcrumbs(defaultItems);

    const homeIcon = screen.getByAltText('home');
    expect(homeIcon).toBeInTheDocument();
    expect(homeIcon).toHaveAttribute('src', homeIconSrc); // Mocked home icon path
    expect(homeIcon).toHaveClass('w-4 h-4 me-2.5 p-0');
  });

  it('matches snapshot for breadcrumbs', () => {
    const items: BreadcrumbItem[] = [
      { label: 'Home', href: ROUTE.ROOT },
      { label: 'Projects', href: ROUTE.PROJECT }
    ];

    const { asFragment } = renderBreadcrumbs(items);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the default "Home" item when no items are provided', () => {
    // Render Breadcrumbs without passing the `items` prop
    render(
      <Router>
        <Breadcrumbs />
      </Router>
    );

    const homeElement = screen.getByText('Home');

    // Check that the default "Home" item is rendered
    expect(homeElement).toBeInTheDocument();
    expect(homeElement.closest('a')).toHaveAttribute('href', ROUTE.ROOT); // Ensures it links to the root
  });
});
