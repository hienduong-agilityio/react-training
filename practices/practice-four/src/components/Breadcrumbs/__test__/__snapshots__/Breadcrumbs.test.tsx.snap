// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Component matches snapshot for breadcrumbs 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
  >
    <ul
      class="flex items-center space-x-1 md:space-x-2 rtl:space-x-reverse"
    >
      <li
        class="inline-flex items-center"
      >
        <a
          class="flex items-center text-sm font-medium text-gray-400 hover:text-primary-500"
          href="/"
        >
          <img
            alt="home"
            class="w-4 h-4 me-2.5 p-0"
            src="homeIcon.svg"
          />
        </a>
      </li>
      <li
        class="inline-flex items-center"
      >
        <img
          alt="Arrow Right"
          class="rtl:rotate-180 w-6 h-6 text-gray-400 mx-1"
          src="arrowRightIcon.svg"
        />
        <a
          class="ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="inline-flex items-center"
      >
        <img
          alt="Arrow Right"
          class="rtl:rotate-180 w-6 h-6 text-gray-400 mx-1"
          src="arrowRightIcon.svg"
        />
        <a
          class="ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2"
          href="/projects"
        >
          Projects
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
